### PR TITLE
perf(tree): re-use intermediate nodes

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

Computation of overlay root with cached intermediate nodes. Will likely be broken down into smaller PRs.

### CODE REVIEW

**Review by Rjected — APPROVED** [review]
lgtm

**Review by gakonst — APPROVED** [review]

**Review by mattsse — APPROVED** [review]

File: crates/chain-state/src/memory_overlay.rs [hunk 1]
```diff
@@ -22,10 +24,12 @@ impl MemoryOverlayStateProvider {
     /// Create new memory overlay state provider.
     pub fn new(in_memory: Vec<ExecutedBlock>, historical: Box<dyn StateProvider>) -> Self {
         let mut hashed_post_state = HashedPostState::default();
+        let mut trie_updates = TrieUpdates::default();
         for block in &in_memory {
             hashed_post_state.extend(block.hashed_state.as_ref().clone());
+            trie_updates.extend(block.trie.as_ref().clone());
         }
```
[the remark this answers is not in the document]
Comment (emhane) [reply crates/chain-state/src/memory_overlay.rs:31]: we can use `Arc` around the trie update and hashed post state in the `ExecutedBlock` type

File: crates/storage/provider/src/providers/bundle_state_provider.rs [hunk 2]
```diff
@@ -99,6 +110,16 @@ impl<SP: StateProvider, EDP: ExecutionDataProvider> StateRootProvider
         self.state_provider.hashed_state_root_with_updates(state)
     }
 
+    fn hashed_state_root_from_nodes_with_updates(
+        &self,
+        _nodes: TrieUpdates,
+        _hashed_state: HashedPostState,
+        _prefix_sets: TriePrefixSetsMut,
+    ) -> ProviderResult<(B256, TrieUpdates)> {
+        // TODO:
```
Comment (mattsse) [inline crates/storage/provider/src/providers/bundle_state_provider.rs:119]: can this be implemented?

File: crates/trie/db/src/state.rs [hunk 3]
```diff
@@ -92,29 +93,37 @@ pub trait DatabaseStateRoot<'a, TX>: Sized {
...
+
+    /// TODO:
+    fn overlay_root_from_nodes(
+        tx: &'a TX,
+        intermediate_nodes: TrieUpdates,
+        post_state: HashedPostState,
+        prefix_sets: TriePrefixSetsMut,
+    ) -> Result<B256, StateRootError>;
+
+    /// TODO:
```
Comment (mattsse) [inline crates/trie/db/src/state.rs:121]: what are the todos here?

**Review by mattsse — COMMENTED** [review]
only have questions re todos

File: crates/trie/db/src/state.rs (hunk 3 above)
Comment (rkrasiuk) [reply crates/trie/db/src/state.rs:121]: missing comments

**Review by mattsse — APPROVED** [review]

**Pull request merged**